src/Nethermind.Verkle/Proofs/MultiProof.cs

```diff
@@ -118,20 +118,38 @@ public bool CheckMultiProof(Transcript transcript, VerkleVerifierQuery[] queries
         transcript.AppendPoint(d, "D");
         FrE t = transcript.ChallengeScalar("t");
 
-        FrE[] g2Den = new FrE[queries.Length];
+        FrE[] groupedEvals = new FrE[domainSize];
+        // Calculate groupedEvals = r * y_i.
         for (int i = 0; i < queries.Length; i++)
         {
-            g2Den[i] = t - FrE.SetElement(u0: queries[i].ChildIndex);
+            groupedEvals[queries[i].ChildIndex] += powersOfR[i] * queries[i].ChildHash;
         }
-        g2Den = FrE.MultiInverse(g2Den);
+
+
+        // REMOVABLE COMMENT: The idea here is to avoid doing a lot of repeated inverses. If we have thousands of 
+        // queries to check, there can only be up to 256 evaluations points.
+        FrE[] helperScalarDens = new FrE[domainSize];
+        // Compute helperScalarsDen = 1 / (t - z_i).
+        for (int i = 0; i < domainSize; i++)
+        {
+            helperScalarDens[i] = t - FrE.SetElement(u0: queries[i].ChildIndex);
+        }
+        helperScalarDens = FrE.MultiInverse(helperScalarDens);
```

Comment:
Note this switched from `queries.Length` to `domainSize`, which is where we save work.
Note that despite batching inverses was doing 1 inverse, the Montgomery trick still implies `O(n)` multiplications that aren't happening now.